Eastern-Research-Group/mywaterway
# Feature/migrate to typescript

## Related Issues:
* N/A (for now – @cschwinderg feel free to make an issue in Jira if you'd like... or I can do so next week)

## Main Changes:
* Adds TypeScript to client app, and begins the gradual migration from Flow.
* Adds other packages/scripts included in newer `create-react-app` bootstrapped projects (e.g. `testing-library` packages, `web-vitals`, removal of `service-worker` script...which we weren't using anyway)
* Converts a single file to TypeScript – the `FetchedData` context component. We'll want to gradually rename files from `.js` to `.ts` and `.tsx` but I'm holding off on changing more files in this PR until we can do it in a coordinated way to prevent merge conflicts from existing branches/PRs.

## Steps To Test:
1. Normally you'd just `npm ci` from the client app directory like normal for any new/updated packages...and that's still probably all you need to do. I replaced the `npm-shrinkwrap.json` file with the more commonly used `package-lock.json` file (we're not publishing the app to a package registry, so a shrinkwrap file isn't needed). The package-lock file is checked into the repo (as usual), but to play it safe and ensure your npm cache is cleared, it might be best to first manually delete the client app's `node_modules` directory and then run `npm ci` from the client app (the `npm ci` script does that, so you may be fine just doing that as normal).
2. Run the app – all should be the same, but please let me know if you run into any issues!

## TODO:
- [ ] Gradually covert JS files to TS/TSX files in a coordinated manner – there will be lots of TS errors we'll need to fix along the way, which will only make the codebase more robust.
